File: packages/nitro-renderer/src/avatar/alias/AssetAliasCollection.ts
import type { IGraphicAsset } from '@nitrodevco/nitro-api';

import { GetAssetManager } from '../../assets';
import { AssetAlias } from './AssetAlias';

export class AssetAliasCollection {
    private _aliases: Map<string, AssetAlias> = new Map();

    public dispose(): void {
        this._aliases = new Map();
    }

    public reset(): void {
        this.init();
    }

    public init(): void {
        for (const collection of GetAssetManager().collections.values()) {
            if (!collection) continue;

            const aliases = collection.data && collection.data.aliases;

            if (!aliases) continue;

            for (const alias of aliases) {
                const assetAlias = new AssetAlias(alias);

                this._aliases.set(assetAlias.name, assetAlias);
            }
        }
    }

    public hasAlias(name: string): boolean {
        return !!this._aliases.get(name);
    }

    public getAssetName(name: string): string {
        let assetName = name;
        let levels = 5;

        while (this.hasAlias(assetName) && (levels >= 0)) {
            const alias = this._aliases.get(assetName);

            if (!alias || !alias.link) break;

            assetName = alias.link;
            levels--;
        }

        return assetName;
    }

    public getAsset(name: string): IGraphicAsset | undefined {
        return GetAssetManager().getAsset(this.getAssetName(name));
    }
}

File: packages/nitro-swift/Sources/NitroAvatar/AssetAliasCollection.swift
import Foundation
import SpriteKit

import NitroAssets

/// Swift port of `AssetAlias` (packages/nitro-renderer/src/avatar/alias/AssetAlias.ts).
public struct AssetAlias {
    public let name: String
    public let link: String
    public let flipH: Bool
    public let flipV: Bool

    public init(_ data: AssetAliasDefinition) {
        name = data.name ?? ""
        link = data.link ?? ""
        flipH = data.flipH ?? false
        flipV = data.flipV ?? false
    }
}

/// Swift port of `AssetAliasCollection` (packages/nitro-renderer/src/avatar/alias/AssetAliasCollection.ts).
/// Avatar asset lookups go through this instead of `AssetManager.getAsset` directly, so a bundle's
/// `aliases` (`IAssetData.aliases`, e.g. one figure part reusing another's texture) get resolved.
public final class AssetAliasCollection {
    private var aliases: [String: AssetAlias] = [:]
    private let assetManager: AssetManager

    public init(assetManager: AssetManager) {
        self.assetManager = assetManager
    }

    /// Rebuilds the alias table from every currently-loaded collection. Call again after loading
    /// more `.nitro` bundles.
    public func reset() {
        aliases.removeAll()

        for collection in assetManager.collections.values {
            for aliasData in collection.data.aliases ?? [] {
                let alias = AssetAlias(aliasData)

                aliases[alias.name] = alias
            }
        }
    }

    public func hasAlias(_ name: String) -> Bool { aliases[name] != nil }

    public func getAssetName(_ name: String) -> String {
        var assetName = name
        var levels = 5

        while hasAlias(assetName), levels >= 0 {
            guard let alias = aliases[assetName], !alias.link.isEmpty else { break }

            assetName = alias.link
            levels -= 1
        }

        return assetName
    }

    public func getAsset(_ name: String) -> GraphicAsset? {
        assetManager.getAsset(getAssetName(name))
    }
}
